src/mon/commands/Command.h

```diff
@@ -0,0 +1,333 @@
+    stringstream ss;
+    bufferlist rdata;
+
+    string format;
+    cmd_getval(cct, cmdmap, "format", format, string("plain"));
+    FormatterRef f(Formatter::create(format));
+
+    string prefix;
+    cmd_getval(cct, cmdmap, "prefix", prefix);
+
+    return do_prepare(op, prefix, cmdmap,
+	ss, rdata, f, pending_map, stable_map);
+  }
+
+  virtual bool do_preprocess(
+      MonOpRequestRef op,
+      const string &prefix,
+      const cmdmap_t &cmdmap,
+      stringstream &ss,
+      bufferlist rdata,
+      FormatterRef f,
+      const Stable &stable_map) = 0;
+
+  virtual bool do_prepare(
+      MonOpRequestRef op,
+      const string &prefix,
+      const cmdmap_t &cmdmap,
+      stringstream &ss,
+      bufferlist rdata,
+      FormatterRef f,
+      Pending &pending_map,
+      Stable &stable_map) = 0;
+
+  virtual bool handles_command(const string &prefix) = 0;
+
+ private:
+  Command() = delete;
+};
+
+
+/* ReadCommand
+ *
+ * Defines a templated Read-only command. These will be called during
+ * preprocess. This class abstracts a derived class' command definition, by
+ * redefining the preprocess command from the 'Command' class, and then
+ * calling a function to be implemented by the class deriving from this one.
+ *
+ * Preparing is still left defined as the 'Command' class has it defined,
+ * which is basically returning 'false'.
+ */
+template<typename T, typename Stable, typename Pending>
+struct ReadCommand : public Command<T, Stable, Pending>
+{
+
+  explicit ReadCommand(
+      Monitor *_mon,
+      PaxosService *_svc,
+      CephContext *_cct) :
+    Command<T, Stable, Pending>(_mon, _svc, _cct)
+  { };
+  virtual ~ReadCommand() { }
+
+  using Command<T, Stable, Pending>::prepare;
+  virtual bool prepare(
+      MonOpRequestRef op,
+      const cmdmap_t &cmdmap,
+      Pending &pending_map,
+      Stable &stable_map) final {
+    ceph_assert(0 == "read-only commands do not perform prepare phase");
+    return false;
+  }
+
+  virtual bool do_prepare(
+      MonOpRequestRef op,
+      const string &prefix,
+      const cmdmap_t &cmdmap,
+      stringstream &ss,
+      bufferlist rdata,
+      FormatterRef f,
+      Pending &pending_map,
+      Stable &stable_map) final {
+    ceph_assert(0 == "read-only commands do not perform prepare phase");
+    return false;
+  }
+
+ private:
+  ReadCommand() = delete;
+
+};
+
+/* WriteCommand
+ *
+ * Defines a templated a Write command. These will be called during prepare.
+ * This class abstracts a derived class' command definition for the prepare
+ * stage, redefining prepare function inherited from the 'Command' class.
+ *
+ * For commands that will have both the preprocess and prepare stages, said
+ * command implementation will require to derive from both the 'ReadCommand'
+ * and 'WriteCommand' classes, and implement both 'do_{preprocess,prepare}'
+ * functions.
+ */
+template<typename T, typename Stable, typename Pending>
+struct WriteCommand : public Command<T, Stable, Pending>
```

Comment:
can we mark `WriteCommand ` `final`? as it is not supposed to be inherited, i guess?

Reply:
both `WriteCommand` and `ReadCommand` can (and I'd argue _should_) be inherited for clarity by each service (see specific services headers). Also, I think you're looking at single commits rather than the final thing, because this is an older version of the code.

Reply:
+1 to @jecluis' comment I like having the explicit types for derivation. (It would be cool to find a way to detect that at compile-time, but having the proper relationship be clear will be a huge improvement.)